Add -Wredundant-decls to default CFLAGS

mkannwischer opened the pull request:
See individual commit messages.

hanno-becker commented on scripts/autogen:
```diff
@@ -605,31 +605,34 @@ def gen_monolithic_undef_all_core(filt=None):
+    def randombytes(c):
+        return "randombytes.h" in c
+
+    def shared_headers(c):
```
This is misnamed, we also consider `*.c` files

mkannwischer replied:
fixed.

hanno-becker reviewed (changes requested):
Thanks @mkannwischer, this LGTM!

Two small points:
* A misnamed function in `autogen`
* A few typos in commit messages ("a one" -> "one")
*

hanno-becker commented on .github/workflows/ci.yml:
```diff
@@ -402,6 +402,38 @@ jobs:
           acvp: false
+  # ensure that kem.h and mlkem_native.h; api.h and native backends are compatible
+  check-apis:
+    needs: [quickcheck, quickcheck-windows, quickcheck-c90, quickcheck-lib, examples, lint, lint-markdown-link]
```
This can be part of the basic CI tests, I think, so the `needs` line can go.

mkannwischer replied:
fixed.

mkannwischer commented:
I've addressed your comments @hanno-becker. 
I also fixed the problems caused by the rebase on top of da5c19c1521622d0834571706396e9af370f5411. 

Please have another look.
